### PR TITLE
BuildModelFromDAG class for causal components

### PR DESCRIPTION
**Pull request by cetagostini**
## Description

Introduces the `BuildModelFromDAG` class to build PyMC models directly from a causal DAG and tabular data, with validation for dims, coords, and priors. Adds comprehensive tests for parsing, validation, error handling, and model construction, ensuring robust integration and usage.

> [!NOTE]
> In the future, I'm planning to extend the class to get a forward pass and then be able to apply saturation and lagging for certain arrows in the dag.

[Example notebook here, available for PyMC People](https://colab.research.google.com/drive/1ul03qZlxnOAqK0ENSqo0LsJMPXm8KHoz#scrollTo=QXbJ86_X_VOj).

To-do's:
- [x] Allow for intercept toggle when we are doing time series (e.g: `intercept=Prior(. . .. )`).

## Related Issue

- [ ] Closes #
- [x] Related to https://github.com/pymc-devs/pymc-extras/issues/572#issuecomment-3238314484

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://www.pymc-marketing.io/en/latest/contributing/index.html). Feel free to comment [`pre-commit.ci autofix` to auto-fix](https://pre-commit.ci/#configuration-autofix_prs).
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks) using [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

----
📚 Documentation preview 📚: https://pymc-marketing--1950.org.readthedocs.build/en/1950/

### CODE REVIEW

File: pymc_marketing/mmm/causal.py [hunk 1]
```diff
@@ -34,6 +45,428 @@ def __init__(self, *args, **kwargs):
...
+            "slope": Prior("Normal", mu=0, sigma=1, dims=slope_dims),
+            "likelihood": Prior(
+                "Normal",
+                sigma=Prior("HalfNormal", sigma=1),
+                dims=self.dims,
+            ),
+        }
+
+    @staticmethod
+    def _parse_dag(dag_str: str) -> nx.DiGraph:
```
Comment (williambdean) [inline pymc_marketing/mmm/causal.py:208]: Is this logic that we have to maintain? Can a digraph just be passed into the initializer?

File: pymc_marketing/mmm/causal.py [hunk 2]
```diff
@@ -34,6 +45,428 @@ def __init__(self, *args, **kwargs):
...
+            if not isinstance(prior, Prior):
+                continue
+            for d in _to_tuple(getattr(prior, "dims", None)):
+                if d not in self.coords:
+                    raise ValueError(
+                        f"Dim '{d}' declared in Prior '{prior_name}' must be present in coords."
+                    )
+
+        # 4) Enforce that likelihood dims match class dims exactly
+        likelihood_prior = self.model_config.get("likelihood")
```
Comment (williambdean) [inline pymc_marketing/mmm/causal.py]: What happens if there is no likelihood key?

Comment (cetagostini) [reply pymc_marketing/mmm/causal.py]: That will never happen. We handle that under the hood.

File: pymc_marketing/mmm/causal.py (hunk 1 above)
Comment (cetagostini) [reply pymc_marketing/mmm/causal.py:208]: Let me take a look.

File: pymc_marketing/mmm/causal.py (hunk 2 above)
Comment (williambdean) [reply pymc_marketing/mmm/causal.py]: don't use get then

File: pymc_marketing/mmm/causal.py (hunk 1 above)
Comment (cetagostini) [reply pymc_marketing/mmm/causal.py:208]: Better to keep it, trying to do it with native nx.digraph and they usually return multidigraphs, and makes everything a bit messy. I can make an issue for the future to delegate more into this, but currently just looks a bit of extra work with not direct benefit from keeping this few lines of code!

File: pymc_marketing/mmm/causal.py (hunk 2 above)
Comment (cetagostini) [reply pymc_marketing/mmm/causal.py]: Done!

File: tests/mmm/test_causal.py [hunk 3]
```diff
@@ -714,3 +743,76 @@ def test_compute_adjustment_sets(
...
+    monkeypatch.setattr("pymc_marketing.mmm.causal.nx", None)
+
+    with pytest.raises(
+        ImportError,
+        match=(
+            r"To use Causal Graph functionality, please install the "
+            r"optional dependencies with: pip install pymc-marketing\[dag\]"
+        ),
+    ):
+        BuildModelFromDAG._parse_dag("A->B")
```
Comment (williambdean) [inline tests/mmm/test_causal.py:785]: no need to test hidden methods

File: tests/mmm/test_causal.py [hunk 4]
```diff
@@ -475,6 +475,35 @@ def test_likelihood_dims_none_init_ok(causal_df):
...
+    builder.model_config = {
+        "slope": Prior("Normal", mu=0, sigma=1),
+        "likelihood": Prior(
+            "Normal", sigma=Prior("HalfNormal", sigma=1), dims=("date",)
+        ),
+    }
+
+    # This should raise the specific ValueError
+    with pytest.raises(ValueError, match=r"'coords' is required and cannot be None\."):
+        builder._validate_coords_required_are_consistent()
```
Comment (williambdean) [inline tests/mmm/test_causal.py:525]: No need to test pydantic behavior

File: tests/mmm/test_causal.py (hunk 3 above)
Comment (cetagostini) [reply tests/mmm/test_causal.py:785]: If not codecov complains quite a bit

File: tests/mmm/test_causal.py (hunk 4 above)
Comment (cetagostini) [reply tests/mmm/test_causal.py:525]: This checks `_validate_coords_required_are_consistent` function. Not following the comment here, but codecov was asking for it.

File: pymc_marketing/mmm/causal.py [hunk 5]
```diff
@@ -34,6 +45,476 @@ def __init__(self, *args, **kwargs):
...
+
+    Examples
+    --------
+    Minimal example using DOT format:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import numpy as np
+        from pymc_marketing.mmm.causal import BuildModelFromDAG
```
Comment (juanitorduz) [inline pymc_marketing/mmm/causal.py]: Can we isort here ? :)

File: pymc_marketing/mmm/causal.py [hunk 6]
```diff
@@ -34,6 +45,476 @@ def __init__(self, *args, **kwargs):
...
+
+        # Validate coords are present and consistent with dims, priors, and df
+        self._validate_coords_required_are_consistent()
+
+        # Validate prior dims consistency early (does not require building the model)
+        self._warning_if_slope_dims_dont_match_likelihood_dims()
+        self._validate_intercept_dims_match_slope_dims()
+
+    @property
+    def default_model_config(self) -> dict:
```
Comment (juanitorduz) [inline pymc_marketing/mmm/causal.py]: can we add the explicit type `dict[str, Prior]` ?

File: pymc_marketing/mmm/causal.py (hunk 1 above)
Comment (juanitorduz) [reply pymc_marketing/mmm/causal.py:208]: We then need to write tests for edge cases :)
